Postage stamp cutouts with WCS

larrybradley opened the pull request:
This PR is an implementation of https://github.com/astropy/astropy/issues/3594.  I still need to add tests.

larrybradley commented:
This PR is ready for review.

astrofrog commented:
Thanks! Before I review this, I was wondering whether it would make sense to also add a section in the actual documentation? (and minor comment: needs a changelog entry)

pllim commented:
:+1: on adding actual documentation in `docs`. As a user that does not really know what this is supposed to do, I would want to see example with screenshots (or `plt.imshow()` results) in the actual doc.

larrybradley commented:
@astrofrog  I've added the changelog entry and I will add a section in the documentation.

larrybradley commented:
I added a new documentation page for `nddata.utils`, which includes the `Cutout` documentation.

This is ready for review again.

mcara commented:
Besides the suggestion on stripping out look-up table distortions from the WCS (see #3594), Iwould like to suggest that this function also updates LTV/LTM so that:
1. reproduce the behavior of IRAF's 'imcopy' task;
2. when a cutout image is loaded in DS9, DS9 will display "physical" coordinates.

adonath commented:
It seems to me, that this cutout class is a basic implemenation of bounding box functionality, that already was requested several times (see e.g. https://github.com/astropy/imageutils/issues/4).  
This functionallity is also needed again in @patti's GSoC project, but with a few additional features.
(see api proposal here: https://github.com/patti/astropy-api/blob/master/model_discretization_api.py, and the ´NDWCSCutoutMixin´ at the bottom). So wouldn't it make sense to tackle this in more general way (not only in the context of WCS stamp cutouts)? My concrete suggestion would be to implement this as a mixin class for NDData, that adds cutout / boundig box functionlity to NDData objects (see api-proposal, that I've linked above). Any thoughts on this? @cdeil and @patti

cdeil commented:
I haven't read this PR and don't have time now, so here's just a general comment: this PR partly overlaps with @patti's GSoC work she has started in #3909. I think it might be great if this could be merged very soon and then @patti or anyone can continue improving it in new PRs.
There's many slightly different use cases for cutouts, and we'll probably never build a perfect system (easy to use, efficient, flexible), so it's best to have something and make the use cases we are a aware of work.
And the 1.1 release is not imminent ... there is some time to adjust and improve even based on user feedback.

So @larrybradley – as far as you're concerned, is this ready for final review or did you want to do more work here, as suggested by @mcara  or @adonath ?

larrybradley commented:
As far as I'm concerned, this can be merged now and improved later.  I want to strip the table distortions, if present (but I don't think that will be a common case) and update the `LTV` keywords (not necessary, but a convenience), but I can add those in a separate PR.  I actually need something like this `Cutout` class (with the slices, bboxes, and origins attributes in both the large and small images, properly taking the boundaries into account) instead of an `NDData`-based object like @adonath suggested.

Please feel free to review and merge now.

mwcraig commented:
I'll review Saturday at the latest -- @astrofrog or @eteq please speak up if you would also like to review before merging (this is labelled as nddata, wcs and coordinates).

astrofrog commented:
I won't have a chance to review this in the next week, but if @mwcraig and @cdeil both review this and are happy with it, then it should be fine to merge :)

cdeil commented on astropy/nddata/tests/test_utils.py:
```diff
@@ -316,3 +320,87 @@ def test_block_size_len(self):
+
+
+class TestCutout(object):
+    data = np.arange(20.).reshape(5, 4)
```
I think I've never seen this way to set up test data ... is it OK? Is there a possibility that one test modifies this (explicitly or by accident) and then the tests are connected?

I think the recommended practice is fixtures or `setup_class` classmethods?
http://pytest.org/latest/xunit_setup.html#class-level-setup-teardown
https://github.com/astropy/astropy/blob/5d419bbc438701e3bc99819ae5a36d9de95f5d41/astropy/vo/validator/tests/test_inpect.py#L21

astrofrog replied:
Yes, best use ``setup_class``

larrybradley replied:
OK, will do.

larrybradley replied:
Done.

cdeil commented on docs/nddata/utils.rst:
```diff
@@ -0,0 +1,257 @@
+for array operations.
+
+
+Cutout Images
```
Can you please link back from the `Cutout` class docstring to this high-level docs section?
Makes it easy to learn how it works and go back and forth in the HTML docs.

larrybradley replied:
Done.

cdeil commented on docs/nddata/utils.rst:
```diff
@@ -0,0 +1,257 @@
+Getting Started
+^^^^^^^^^^^^^^^
+
+The `~astropy.nddata.utils.Cutout` class can be used to create a
```
This `Cutout` class only works with 2D images (I think), but is located in `nddata`.
Probably it's not easy to make it work nicely or at all for ND?
So maybe leave implementation as-is, but rename to `CutoutImage` or `Cutout2D`?
(so that in the future someone can implement the equivalent 1D or 3D or ND functionality ...

larrybradley replied:
I renamed the class to `Cutout2D`.  Thanks.

cdeil commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,267 @@ def block_replicate(data, block_size, conserve_sum=True):
+            The corresponding ``(y, x)`` pixel position in the large
+            array.
+        """
+        return tuple(position[i] + self.origin_large[i] for i in [0, 1])
```
The `to_large` and `from_large` one-line methods are the only ones not covered by unit tests.
I think they are covered by docs tests, but those don't run for Python 3, so maybe useful to add a unit test for those?

larrybradley replied:
Done.

cdeil commented on docs/nddata/utils.rst:
```diff
@@ -0,0 +1,257 @@
+    plt.imshow(data, origin='lower')
+
+Now let's create a cutout array for the single object in this image.
+We create a cutout array centered at position ``(100.1, 49.7)`` (``(y,
```
I find this hard to read:
![image](https://cloud.githubusercontent.com/assets/852409/8621876/59a4fdd2-2728-11e5-91fc-05363345946f.png)

Maybe instead write it with an equals sign?
```
(y, x) = (100.1, 49.7)
```

larrybradley replied:
Done.

cdeil commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,267 @@ def block_replicate(data, block_size, conserve_sum=True):
+        Create a cutout object from a 2D array.
+
+        The returned object will contain the 2D cutout array and the
+        updated WCS, if a `~astropy.wcs.WCS` object is input.
```
Can you add a sentence or two describing whether this creates a view into `data` or makes a copy of part of it?
I.e. does the cutout object have references back to `data` (or `wcs`) and can / should I use it to modify that part of the `data` array?

One application @adonath and I have for cutouts is that we have a survey image.
Then we create cutouts of overlapping groups of sources (currently shelling out and calling `ftcopy` to make sure the FITS header is updated correctly :disappointed: ) and fit models on those.
Then  at the end we copy the small model images back into a survey model image.

I guess my question is: is the copying back into the survey image possible / easy with what's implemented here?

larrybradley replied:
`cutout.data` is a view of the original large `data` array.  `cutout.wcs` is a copy of the original `wcs` with the updated `CRPIX` values.  I've updated the docs.

larrybradley replied:
I also added a `copy` option.  If `True` then `Cutout2D.data` will be a copy instead of a view.

cdeil commented on docs/nddata/utils.rst:
```diff
@@ -0,0 +1,257 @@
+    wcs.wcs.crval = [position.ra.value, position.dec.value]
+    wcs.wcs.crpix = [100, 50]
+    cutout = Cutout(data, position, (30, 40), wcs=wcs)
+    plt.imshow(cutout.data)
```
Add `origin='lower'` here (currently the image is flipped in y wrt. rest of the tutorial).

larrybradley replied:
Done.

cdeil commented:
I gave this a quick review.
Sorry, I don't have time to try this out.

:+1: to merge this soon ... I want to use this ... thanks @larrybradley !

mwcraig commented:
I ran into a problem with this image: https://drive.google.com/file/d/0B8BllHJSoO-kYXdtUGdCRkpKNHM/view?usp=sharing

mwcraig commented:
@larrybradley -- any updates? Let me know if you want me to try to chase down the problem with the image linked to above.

larrybradley commented:
@mwcraig I'll look into your issue tomorrow, thanks.

astrofrog commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,270 @@ def block_replicate(data, block_size, conserve_sum=True):
+        else:
+            self.wcs = None
+
+    def to_large(self, position):
```
This method name isn't very explicit, and I am wondering whether we can do better? What about ``to_original_position`` or something like this?

larrybradley replied:
Thanks, @astrofrog.  The methods are now `to_original_position()` and `to_cutout_position()`.  I also renamed everything from "large" to "original" and "small" to "cutout".

larrybradley commented:
@mwcraig I fixed the issue you were seeing.  All of `Cutout2D`'s positions are now in `(x, y)` order.  Here's an example using your FITS image:  https://gist.github.com/larrybradley/36543ff8b54551752fc1.

I also added a `plot_on_original()` method to plot the cutout bounding box on the original image (example in the above gist).

cdeil commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,322 @@ def block_replicate(data, block_size, conserve_sum=True):
+class Cutout2D(object):
+    def __init__(self, data, position, shape, wcs=None, mode='trim',
+                 fill_value=np.nan, copy=False):
+        """
```
I think docstrings are support to be at the class level, not in `__init__`?

larrybradley replied:
They work fine in ``__init__``.  I thought it could be either/or and I prefer have the parameters in ``__init__``.  I moved the one-line docstring to the class level (but it also worked fine in `__init__`).  I can also move the description up to class level if you want.

cdeil commented:
I think this is getting close to merge.
@patti @adonath – Do you want to review this to make sure it's consistent with what you do in #3909 ?
Alternatively, @larrybradley, do you have a time to look there?
If possible, terminology and e.g. bounding box axis order should be consistent.

cdeil commented on astropy/nddata/tests/test_utils.py:
```diff
@@ -316,3 +320,116 @@ def test_block_size_len(self):
+                                                     self.wcs)
+        skycoord_cutout = self.position.from_pixel(c.center_cutout[1],
+                                                   c.center_cutout[0], c.wcs)
+        assert_allclose(skycoord_original.ra.value, skycoord_cutout.ra.value)
```
Would it make sense to use [astropy.tests.helper.assert_quantity_allclose](http://astropy.readthedocs.org/en/latest/api/astropy.tests.helper.assert_quantity_allclose.html) for all these angle (i.e. quantity sub-class) asserts?

larrybradley replied:
Done, thanks.

cdeil commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,322 @@ def block_replicate(data, block_size, conserve_sum=True):
+        return tuple(original_position[i] - self.origin_original[i]
+                     for i in [0, 1])
+
+    def plot_on_original(self, ax=None, fill=False, **kwargs):
```
Maybe return `ax` to make it simpler for the user to continue drawing on that `ax` in case it was constructed in the method?

larrybradley replied:
Done.

mwcraig commented:
@larrybradley  -- looking good! I'll do a more thorough code read-through tonight.

@cdeil -- my inclination would be to merge this and have @patti and @adonath revise with a separate PR if necessary unless @larrybradley really, really wants to make more modifications...

mwcraig commented:
Sorry for the delay -- I've looked through this and had no comments beyond what @cdeil has made.

This looks very useful, and thanks for clarifying/cleaning up some of the docs for `extract_array`, etc!

larrybradley commented:
Thanks, @cdeil and @mwcraig.  I'll finish this tomorrow.

patti commented:
@cdeil, @adonath, @larrybradley

The `Cutout2D` bounding boxes aren't consistent with `Model.bounding_box` in #3909.

**TLDR**
I think it makes the most sense to change how `Cutout2D.bbox_original` and `Cutout2D.bbox_cutout` are defined as opposed to changing `Model.bounding_box`. I volunteer to do this in a separate PR as @mwcraig suggested, especially if it'll help my PR get merged before the end of GSoC :)

___

In 2D:

    Model.bounding_box = ((xmin, xmax), (ymin, ymax))
    Cutout2D.bbox_original = (ymin, ymax, xmin, xmax)


`Cutout2D` is strictly 2D. A model can occupy N-dimensions. For a 3D model, under the `Cutout2D` bounding box convention, 

    Model.bounding_box  =  (ymin, xmin, zmin, ymax, xmax, zmax)

which is not intuitive and kind of ugly.

I thought I might introduce a new property called `bbox_original` that matches the same property in `Cutout2D`. It would be one in the same as `bounding_box` except for the format it accepts and returns.  

    >>> model.bounding_box = ((1, 2), (3, 4))
    >>> model.bounding_box
    ((1, 2), (3, 4))
    >>> model.bbox_original 
    (1, 3, 2, 4)
    >>> model.bbox_original = (4, 5, 6, 7)
    >>> model.bbox_original
    (4, 5, 6, 7)
    >>> model.bounding_box
    ((4, 6), (5, 7))

But this is getting messy. It would probably have to go in `Model` or `Fittable2DModel` and I don't think this would fly with the powers-that-be

Looking at the code, @larrybradley correct me if I'm wrong, it seems like it would be straightforward to change the format of `Cutout2D.bbox_original` and `Cutout2D.bbox_cutout` without introducing any major complications. If so, and there aren't objections, I volunteer to do the work in a separate PR.

mwcraig commented:
@patti -- finally took a quick look at #3909 -- you have been busy!

Is the issue here, in a nutshell, that a tuple of tuples generalizes nicely to higher dimensions than two but a flat tuple doesn't? That makes sense to me.

@larrybradley -- any objections to merging this in and then @patti doing a PR to make the bounding box change? I don't have any sense of how this fits into, e.g. `photutils`, but there aren't any issues with backwards compatibility since this is a new feature.

adonath commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,322 @@ def block_replicate(data, block_size, conserve_sum=True):
+         [ nan   4.   5.]]
+        """
+
+        if mode not in ['partial', 'trim', 'strict']:
```
Isn't that caught in `overlap_slices` and `extract_array`?

larrybradley replied:
Yes, it is caught now in both `overlap_slices` and `extract_array`.  `extract_array` orginally didn't have the `'trim'` option (I added it in this PR) when I wrote that check.  I removed the now redundant `mode` check in `Cutout2D`.  Thanks for catching that.

adonath commented on astropy/nddata/utils.py:
```diff
@@ -456,3 +469,322 @@ def block_replicate(data, block_size, conserve_sum=True):
+            required input.
+
+        shape : tuple
+            The shape (``(ny, nx)``) of the cutout array in pixel
```
I'm not sure if this is easily possible, but I would expect if I can pass the position in sky coordinates, I can do the same for `shape` (i.e. give the size of the region I'd like to cut put in deg). I see it might be complicated for general WCS transformations, where projected regions can have arbitrary shapes...

larrybradley replied:
It's certainly possible.  We do this exact thing for the `photutils` sky apertures.  To prevent duplication, the functions in `photutils.utils.wcs_helpers` would first need to be moved to `astropy`.  I prefer to do this in a separate PR, so I'm going to pass on this for now.

larrybradley commented:
Thanks, @patti.  I modified the `bbox_original` and `bbox_cutout` to match your convention in `Model.bounding_bound`.

larrybradley commented:
@mwcraig Hopefully this is ready to merge now (if all tests pass)!

mwcraig commented:
There are 8 failures against numpy dev, but those seem to be unrelated to this PR and they are occurring in at least a couple other PRs, so merging.

Thanks to @larrybradley for putting this together and to everyone who provided feedback.

ofors commented:
Thanks @larrybradley and @mwcraig : Cutout2D is a really useful class. 
I'm using FITS images with TPV convention and 3rd degree polynomial distortion (PV?_?).
I have two questions:

1) is there a class/method of Cutout2D which returns True/False depending if input (coord,boxsize) overlap with the input image? 
This has been implemented in [astLib.WCS.coordsAreInImage(self, RADeg, decDeg)](astlib.sourceforge.net/docs/astLib/astLib.astWCS.WCS-class.html#coordsAreInImage), and in [def coords_in_image(xc,yc,header,coordsys='celestial'):](https://github.com/keflavich/agpy/blob/master/agpy/get_cutouts.py).
However, I would like to keep using Cutout2D in astropy, because of the availability of improved wcslib5.9, which is crucial for TPV images.

2) which is the schedule for Cutout2D to support TPV images with PV?_? distortions?
I can work on testing side.

larrybradley commented:
@ofors For 1, `Cutout2D` will raise a `NoOverlapError` is there is **no** overlap of the cutout with the input image.  The behavior of partial overlaps can be controlled with the `mode` keyword.  For example, using `mode='strict'` will raise a `PartialOverlapError` if the cutout only partially overlaps the input data.  A partial overlap can also be detected with `mode='trim'`.  In that case the cutout image (e.g. `cutout.data`) will not have the same shape as the requested input `shape`.

A generic function to determine if a `SkyCoord` position is in an image (like `astLib.WCS.coordsAreInImage`) could be useful, but it would belong in `astropy.wcs.utils`.  PRs welcome!

For 2, I've never used TPV images.  `Cutout2D` simply shifts the `CRPIX` values of the input WCS to create the output WCS.  This keeps the tangent point of the projection the same and also works for SIP distortion corrections because `CRPIX` is the reference pixel for those distortions.  If TPV also uses `CRPIX` as the reference pixel, then I would guess that it should work, but I really don't know anything about TPV.  If not, then perhaps someone with TPV experience could make that work in `Cutout2D`.

mwcraig commented:
@ofors -- nothing to add beyond @larrybradley's comments except to clarify that @larrybradley did all the hard work :)

ofors commented:
@larrybradley You're right. I updated my script using Cutout2D and now it's using try & except with mode='partial' or 'trim'. That way, I can tolerate PartialOverlapError exceptions, but I don't cutout the ones throwing NoOverlapError.
I agree a function like coords_in_image would be nice to have (I encourage others to pull a request :). However, right now I managed to live with the exceptions which Cutout2D throws.

As 2, I tried Cutout2D with a TPV P?_? polynomial distorted image and got a 256x256pix cutout. 
The original image is 23.8degx16.0deg FoV, 13.17arcsec/pixel.
The problem with the cutout is that when displaying with ds9 is not showing the correct central target coordinates (ra=311.08,dec=-68.09). Instead it is showing (ra=308.54,dec=-63.43).

> @larrybradley: "For 2, I've never used TPV images. Cutout2D simply shifts the CRPIX values of the 
> input WCS to create the output WCS."

does this mean that Cutout2D doesn't rotate the cutout according the rotation matrix CD[1,2]_[1,2] and it only shifts?
I see @keflavich [cutout.py:L115,116 ](https://github.com/keflavich/agpy/blob/master/agpy/cutout.py) that he is rotating the cutout with CD[1,2]_[1,2].
Does anybody think this could solve my coordinate offset?

larrybradley commented:
@ofors Neither `Cutout2D` or @keflavich's `cutout.py` perform any rotations (why would they?).  `L115:116` of `cutout.py` is simply converting `(xw, yw)` from WCS units into pixel units using the x and y pixel scales (and actually the calculated pixels scales are not correct in `cutout.py` if `CDELT1,2` are not in the header).

Note that `Cutout2D` does not yet allow `shape` to be specified in angular units.  It's on my "to do" list.

I looked at the TPV distortions (http://fits.gsfc.nasa.gov/registry/tpvwcs/tpv.html) with @mcara and because of the extra `r` term, we don't see any easy way to propagate the WCS to the cutout image.  `Cutout2D` works with SIP, so one option would be to convert from TPV to SIP (e.g. http://web.ipac.caltech.edu/staff/shupe/reprints/SIP_to_PV_SPIE2012.pdf).

larrybradley commented:
@ofors `Cutout2D` seems to work fine with TPV images (because like SIP, the reference pixel is also `CRPIX`).  Here's a simple example starting from the FITS file you sent me:

```python
from astropy.io import fits
from astropy.wcs import WCS
from astropy.nddata.utils import Cutout2D
from astropy.wcs.utils import pixel_to_skycoord

f = fits.open('ML0751914_20150701_035326_calib.fits')
data = f[0].data
w = WCS(f[0].header)
cutout = Cutout2D(data, (2000, 2000), (100, 100), wcs=w)

# find the skycoord of a pixel in the cutout image
x_cutout, y_cutout = (5, 10)
skycoord_cutout = pixel_to_skycoord(x_cutout, y_cutout, cutout.wcs)

# now find the skycoord of the corresponding pixel in the original image
x_data, y_data = cutout.to_original_position((x_cutout, y_cutout))
skycoord_original = pixel_to_skycoord(x_data, y_data, w)

print(skycoord_cutout)
<SkyCoord (FK5: equinox=2000.0): (ra, dec) in deg
    (298.10121912, -71.86024496)>

print(skycoord_original)
<SkyCoord (FK5: equinox=2000.0): (ra, dec) in deg
    (298.10121912, -71.86024496)>
```

As you see the `SkyCoord` match.  Perhaps DS9 is giving the wrong coordinates?

larrybradley commented:
I forgot to mention that my example above requires the current development version of `astropy` (post https://github.com/astropy/astropy/pull/3905) to work with FITS headers containing TPV distortions.

ofors commented:
Thanks so much @larrybradley for your code snippet.
Does anybody know how to store the cutout in a FITS file with the original image header and the CRPIX[1,2] values of cutout.wcs?
An example would be most than welcome.

larrybradley commented:
@ofors Please see the documentation for creating a FITS file:
http://astropy.readthedocs.org/en/latest/io/fits/#creating-a-new-fits-file

It is trivial to update a FITS header:
```python
>>> header['CRPIX1'] = cutout.wcs.wcs.crpix[0]
>>> header['CRPIX2'] = cutout.wcs.wcs.crpix[1]
```

Also see the documentation on how to get a FITS header object from a `WCS` object: http://astropy.readthedocs.org/en/latest/api/astropy.wcs.WCS.html#astropy.wcs.WCS.to_header

ofors commented:
Thanks @larrybradley ! That worked fine.
Please let us know when you plan to implement the polynomial distortion of the cutout (not a copy of the original image).

larrybradley commented:
@ofors `Cutout2D` already handles polynomial distortions (e.g. SIP, TPV).  I don't plan on adding anything there.

ofors commented:
Oh! I thought the Warning [here](http://astropy.readthedocs.org/en/latest/api/astropy.nddata.utils.Cutout2D.html) saying : "The cutout WCS object does not currently handle cases where the input WCS object contains tabular distortions." was referring the polynomial distortions P?_?.

larrybradley commented:
No, that refers to the table lookup distortions (i.e. WCS Paper IV).  I'm not planning to handle them anytime soon (if at all).
